impl/src/backend.test.ts

```diff
@@ -0,0 +1,148 @@
+import { fairlyAllocateCredit } from "./backend";
+
+import { strict as assert } from "assert";
+import test from "node:test";
+import * as simplestatistics from "simple-statistics";
+
+interface FairlyAllocateCreditTestCase {
+  name: string;
+  credit: number[];
+  value: number;
+  needsRand?: boolean;
+}
+
+function noRand(): number {
+  throw new Error("no rand expected");
+}
+
+type Interval = [min: number, max: number];
+
+// https://en.wikipedia.org/wiki/Normal_distribution#Quantile_function
+function ppf(q: number, mean: number, stdDev: number): number {
+  return (
+    mean +
+    stdDev * Math.sqrt(2) * simplestatistics.inverseErrorFunction(2 * q - 1)
+  );
+}
+
+function getIntervalApprox(n: number, p: number, alpha: number): Interval {
+  const mean = n * p;
+  const variance = mean * (1 - p);
+  const diff = ppf(1 - alpha / 2, mean, Math.sqrt(variance)) - mean;
+  return [mean - diff, mean + diff];
+}
+
+function getAllIntervals(
+  n: number,
+  creditFractions: readonly number[],
+  alphaTotal: number,
+): Interval[] {
+  const alpha = alphaTotal / creditFractions.length;
+  return creditFractions.map((cf) => getIntervalApprox(n, cf, alpha));
+}
+
+function runFairlyAllocateCreditTest(
+  tc: Readonly<FairlyAllocateCreditTestCase>,
+): void {
+  // TODO: replace with precise sum
+  const sumCredit = tc.credit.reduce((a, b) => a + b, 0);
+  const normalizedFloatCredit = tc.credit.map(
+    (item) => (tc.value * item) / sumCredit,
+  );
+
+  const [rand, k] = tc.needsRand ? [Math.random, 100000] : [noRand, 1];
+
+  const totals = new Array<number>(tc.credit.length).fill(0);
+
+  for (let n = 0; n < k; ++n) {
+    const actualCredit = fairlyAllocateCredit(tc.credit, tc.value, rand);
+
+    assert.equal(actualCredit.length, tc.credit.length);
+
+    for (const [j, actual] of actualCredit.entries()) {
+      assert.ok(Number.isInteger(actual));
+
+      const normalized = normalizedFloatCredit[j]!;
+      const diff = Math.abs(actual - normalized);
+      assert.ok(
+        diff <= 1,
+        `credit error > 1: actual=${actual}, normalized=${normalized}`,
+      );
+
+      totals[j]! += actual / tc.value;
+    }
+
+    assert.equal(
+      // TODO: replace with precise sum
+      actualCredit.reduce((a, b) => a + b, 0),
+      tc.value,
+      `actual credit does not sum to value: ${actualCredit.join(", ")}`,
+    );
+  }
+
+  const intervals = getAllIntervals(
+    k,
+    tc.credit.map((c) => c / sumCredit),
+    0.0001,
+  );
```

Comment:
```suggestion
  // Significance level for the overall test, where each subtest has
  // a null hypothesis that its correct mean is precisely the passed in
  // normalized float credit. This means the test should flake with
  // probability no more than `alphaTotal`.
  const alphaTotal = .0001
  const intervals = getAllIntervals(
    k,
    normalizedFloatCredit,
    alphaTotal
  );
```

Reply:
Maybe the variables aren't named properly, but currently `normalizedFloatCredit` has a `value` factor, whereas, IIUC, `getAllIntervals` must operate on credit values in `[0, 1]`.

Reply:
Oh you're right. I think it is expecting `normalizedFloatCredit - floor(normalizedFloatCredit)` i.e. the fractional part.

Reply:
So that, and I would go further and avoid using the interval test when `k == 1`:

```suggestion
  const alpha = 0.00001; // Probability of test failing at random.
  const intervals = (k > 1) ? getAllIntervals(
    k,
    normalizedFloatCredit.map((c) => c - Math.floor(c)),
    alpha,
  ) : normalizedFloatCredit.map((c) => [c, c]);
```

I would reduce the value of `k` and increase `alpha` given this arrangement.  No need to run 100k times when `alpha` is what governs the probability of failure: increase that by a factor of 10 instead.